engine/src/main/java/io/zeebe/engine/state/variable/DocumentEntryIterator.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.engine.state.variable;
+
+import io.zeebe.msgpack.spec.MsgPackReader;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import org.agrona.DirectBuffer;
+import org.agrona.collections.Int2IntHashMap.EntryIterator;
+import org.agrona.concurrent.UnsafeBuffer;
+
+/**
+ * Iterates of a document by reading the offsets directly from the associated {@code
+ * offsetIterator}. Expected usage only through {@link IndexedDocument#iterator()}.
+ *
+ * <p>Note that keys are expected to be strings, and as such the length is removed from the MsgPack
+ * representation before being added as the name in the {@link DocumentEntry}. String values are
+ * kept as is, as values can be any kind of object.
+ */
+final class DocumentEntryIterator implements Iterator<DocumentEntry> {
+
+  private final MsgPackReader reader;
+  private final DocumentEntry entry = new DocumentEntry();
+  private final DirectBuffer document = new UnsafeBuffer();
+
+  private EntryIterator offsetIterator;
+  private int documentLength;
+
+  DocumentEntryIterator() {
+    this(new MsgPackReader());
+  }
+
+  DocumentEntryIterator(final MsgPackReader reader) {
+    this.reader = reader;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return offsetIterator.hasNext();
+  }
+
+  @Override
+  public DocumentEntry next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException();
+    }
+
+    offsetIterator.next();
+    final int keyOffset = offsetIterator.getIntKey();
+    final int valueOffset = offsetIterator.getIntValue();
+
+    reader.wrap(document, keyOffset, documentLength - keyOffset);
```

Comment:
I'm not sure how these values are selected, but if documentLength is the total length and the key is smaller than the value, then shouldn't it only read the key and not the value as well? Perhaps it doesn't really matter. But I thought something like:
```suggestion
    reader.wrap(document, keyOffset, documentLength - (keyOffset + valueOffset));
```

Reply:
It doesn't really matter, because we use proper read operations afterwards. The important part is that the length is still valid, i.e. length + offset doesn't end up pointing outside of the buffer.